core/src/main/scala/org/apache/spark/sql/delta/util/DeltaFileOperations.scala

```diff
@@ -336,6 +336,9 @@ object DeltaFileOperations extends DeltaLogging {
   def tryDeleteNonRecursive(fs: FileSystem, path: Path, tries: Int = 3): Boolean = {
     try fs.delete(path, false) catch {
       case _: FileNotFoundException => true
+      case e: IOException if tries == 1 =>
```

Comment:
"first" doesn't mean `tries == 1`. This method is called recursively (ironic, given that it deletes non-recursively).
I will be called with tries = N, then N-1, then N-1, ...

So the only way to detect the "first" call is to duplicate the code to another method. The steps will be
a) call the public API `tryDeleteNonRecursive`. If it fails, this is the first attempt.
b) that method will have to call `_tryDeleteNonRecursive`, or some other **private** internal method helper

OR

Another thing we could do is add a default `first: Boolean = true` param to this public API, and then have all recursive calls set `first = false`.

Honestly none of these options seems great. @zsxwing any other ideas?